### PR TITLE
Token issue and transfer transaction validation

### PR DESCRIPTION
**Pull request by Naviabheeman**
Enhanced ColorIdentifier class with comparison operators and new constructors
Updated CScriptCheck class to save the colorId after verifyScript validation. The saved colorId is then used to check token balance.
Added checks to validate colorIdentifier in transactions with OP_COLOR outputs.
Added TokenBalance checks

Added unit tests for ColorIdentifier class
Added unit tests for All types of Tokens - check token issue, transfer, aggregation etc.

### CODE REVIEW

File: src/policy/policy.cpp [hunk 1]
```diff
@@ -51,7 +51,8 @@ CAmount GetDustThreshold(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
 
 bool IsDust(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
 {
-    return (txout.nValue < GetDustThreshold(txout, dustRelayFeeIn));
+    return (txout.nValue < GetDustThreshold(txout, dustRelayFeeIn) 
+    && !(txout.nValue == 1 and GetColorIdFromScript(txout.scriptPubKey).type == TokenTypes::NFT));
```
Comment (azuchi) [inline src/policy/policy.cpp:55]: Since a small number of tokens can be traded outside of the NFT, let's keep the tokens out of the Dust. If so, is there a problem?

File: src/test/txvalidation_tests.cpp [hunk 2]
```diff
@@ -59,5 +61,571 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChainSetup)
...
+tokenIssueTx(from coinbaseSpendTx) - 100 tokens
+tokenTransferTx - 1. no fee
+                - 2. split into 50 + 40 tokens - token balance error
+                - 3. success
+test transaction reissue
+tokenAggregateTx- 1. no fee
+                - 2. add extra tokens - token balance error
+                - 3. success
+TODO: tokenBurnTx     - 1. no fee
+                - 2. add extra TPC - token balance error
```
Comment (azuchi) [inline src/test/txvalidation_tests.cpp:179]: Will this TODO be resolved with another pull request?

File: src/test/txvalidation_tests.cpp [hunk 3]
```diff
@@ -59,5 +61,571 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChainSetup)
...
+    testTx(this, MakeTransactionRef(tokenIssueTx), false, "invalid-colorid");
+}
+
+/*
+Test token type REISSUABLE
+Txs:
+coinbaseSpendTx
+tokenIssueTx(from coinbaseSpendTx) - 100 tokens
+tokenTransferTx - 1. no fee
+                - 2. split into 50 + 40 tokens - token balance error
```
Comment (azuchi) [inline src/test/txvalidation_tests.cpp:172]: Also balance increases case 100 -> 50, 51 is need.

File: src/test/txvalidation_tests.cpp [hunk 4]
```diff
@@ -59,5 +61,571 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChainSetup)
...
+    tokenTransferTx.vout.resize(1);
+    tokenTransferTx.vout[0].nValue = 1;
+
+    Sign(vchSig, key0, tokenIssueTx.vout[0].scriptPubKey, tokenIssueTx, 0, tokenTransferTx, 0);
+    tokenTransferTx.vin[0].scriptSig = CScript() << vchSig << vchPubKey0;
+    Sign(vchSig, coinbaseKey, m_coinbase_txns[3]->vout[0].scriptPubKey, coinbaseIn2, 1, tokenTransferTx, 0);
+    tokenTransferTx.vin[1].scriptSig = CScript() << vchSig;
+
+    testTx(this, MakeTransactionRef(tokenTransferTx), true);
+}
```
Comment (azuchi) [inline src/test/txvalidation_tests.cpp:630]: add tests for the following cases?
- Transfer of multiple tokens
- Issue and transfer tokens at the same tx

File: src/test/txvalidation_tests.cpp (hunk 2 above)
Comment (Naviabheeman) [reply src/test/txvalidation_tests.cpp:179]: I initially thought I can create another PR. Now I have finished testing token burn cases. So I'll just update this PR

File: src/policy/policy.cpp (hunk 1 above)
Comment (Naviabheeman) [reply src/policy/policy.cpp:55]: no, it think I can check the script instead and not check for dust in case it is token.

File: src/test/coins_tests.cpp [hunk 5]
```diff
@@ -479,31 +479,34 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
...
     // Good example
-    CDataStream ss2(ParseHex("8ddf77bbd123008c988f1a4a4de2161e0f50aac7f17e7f9555caa4"), SER_DISK, CLIENT_VERSION);
+    CDataStream ss2(ParseHex("8ddf77bbd123008c988f1a4a4de2161e0f50aac7f17e7f9555caa401"), SER_DISK, CLIENT_VERSION);
     Coin cc2;
     ss2 >> cc2;
     BOOST_CHECK_EQUAL(cc2.fCoinBase, true);
     BOOST_CHECK_EQUAL(cc2.nHeight, 120891U);
     BOOST_CHECK_EQUAL(cc2.out.nValue, 110397);
     BOOST_CHECK_EQUAL(HexStr(cc2.out.scriptPubKey), HexStr(GetScriptForDestination(CKeyID(uint160(ParseHex("8c988f1a4a4de2161e0f50aac7f17e7f9555caa4"))))));
+    BOOST_CHECK_EQUAL(TokenToUint(cc2.type), 1);
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply src/test/coins_tests.cpp:499]: I haven't added the crosscheck between script and token type stored in the coin. I just assumed that when the coin is created in the UTXO db token type would have been validated already. maybe I can optimise it. I'll check.

**Comment by azuchi** [comment]
@Naviabheeman can you rebase to fix conflict?
